blog/_posts/2019-08-01-artifacts.md

```diff
@@ -0,0 +1,179 @@
+
+```julia
+products = [
+    FileProduct("src/data.txt", :data_txt),
+    LibraryProduct("libdataproc", :libdataproc),
+    ExecutableProduct("mungify", :mungify_exe)
+]
+```
+
+With such products defined, the jll package will contain `data_txt`, `libdataproc` and `mungify_exe` symbols exported. For `FileProduct` variables, the exported value is a string pointing to the location of the file on-disk.  For `LibraryProduct` variables, it is a string corresponding to the `SONAME` of the desired library (it will have already been `dlopen()`'ed during the `__init__()` method of the jll package module so typical `ccall()` usage applies), and for `ExecutableProduct` variables, the exported value is a function that can be called to set appropriate environment variables such as `PATH` and `LD_LIBRARY_PATH`.  This is necessary so that nested depdenencies works properly, e.g. `ffmpeg` calling the `x264` binary during video encoding.  Example:
+
+```julia
+using c_simple_jll
+
+# For file products, you can access its file location directly:
+data_lines = open(data_txt, "r") do io
+    readlines(io)
+end
+
+# For library products, you can use the exported variable name in `ccall()` invocations directly
+num_chars = ccall((libdataproc, :count_characters), Cint, (Cstring, Cint), data_lines[1], length(data_lines[1]))
+
+# For executable products, you can use the exported variable name as a function that you can call
+mungify_exe() do mungify_exe_path
+    run(`$mungify_exe_path $num_chars`)
+end
+```
+
+One of the big benefits of this new system is that running binaries that must link against libraries provided by another dependency, or call binaries provided within another dependency all works flawlessly because the wrapper jll packages automatically set up the appropriate environment variables.
+
+## Changes to `build_tarballs.jl`
+
+Not much has changed here, the first is some cleanup of the clunky products API; previously you needed to provide a closure to curry a `prefix` through to your products; this is no longer the case.  To make this explicit, while you would previously write declarations such as:
+
+```julia
+products(prefix) = [
+    LibraryProduct(prefix, "libglib", :libglib)
+]
+```
+
+You now simply write:
+```julia
+products = [
+    LibraryProduct("libglib", :libglib)
+]
+```
+
+The second big change is that dependencies are no longer linked straight to `build.jl` files, rather you give the names of the jll packages you want to install, and they (along with all recursive dependencies) will be installed and symlinked into `${prefix}`, similarly to before.  To make this explicit, while you would previously write declarations such as:
+
+```julia
+dependencies = [
+    "https://github.com/JuliaBinaryWrappers/c_simple_jll.jl/releases/download/c_simple%2Bv1.2.3%2B7/build_c_simple.v1.2.3.jl",
+]
+```
+
+You now simply write:
+```julia
+dependencies = [
+    "c_simple_jll",
```

Comment:
Is there a need for a UUID here?

Reply:
As I was transitioning from the dream state to consciousness this morning, I realized that users will want the ability to put anything they could put into `pkg> add` here.  E.g. they'll want to be able to say `c_simple_jll@v1.2.3`, and also `https://privategitlabinstance/my/private/project_jll.git@v1.2.3`.  So perhaps the only foolproof way to deal with this is to have `dependencies` support not only strings (that must be parsed) but also `PackageSpecs`.